src/mongo/db/exec/shard_namer_impl.h

```diff
@@ -0,0 +1,52 @@
+/**
+ *    Copyright (C) 2019-present MongoDB, Inc.
+ *
+ *    This program is free software: you can redistribute it and/or modify
+ *    it under the terms of the Server Side Public License, version 1,
+ *    as published by MongoDB, Inc.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    Server Side Public License for more details.
+ *
+ *    You should have received a copy of the Server Side Public License
+ *    along with this program. If not, see
+ *    <http://www.mongodb.com/licensing/server-side-public-license>.
+ *
+ *    As a special exception, the copyright holders give permission to link the
+ *    code of portions of this program with the OpenSSL library under certain
+ *    conditions as described in each individual source file and distribute
+ *    linked combinations including the program with the OpenSSL library. You
+ *    must comply with the Server Side Public License in all respects for
+ *    all of the code used other than as permitted herein. If you modify file(s)
+ *    with this exception, you may extend this exception to your version of the
+ *    file(s), but you are not obligated to do so. If you do not wish to do so,
+ *    delete this exception statement from your version. If you delete this
+ *    exception statement from all source files in the program, then also delete
+ *    it in the license file.
+ */
+
+#pragma once
+
+#include "mongo/db/exec/shard_namer.h"
+#include "mongo/db/matcher/matchable.h"
+#include "mongo/db/s/scoped_collection_metadata.h"
+
+namespace mongo {
+
+class ShardNamerImpl : public ShardNamer {
+public:
+    ShardNamerImpl(ScopedCollectionMetadata md);
+
+    const StringData shardName() const override;
+
+    bool isCollectionSharded() const override {
+        return _metadata->isSharded();
+    }
+
+private:
+    ScopedCollectionMetadata _metadata;
```

Comment:
Another reason to avoid the `ShardNamer` interface is that it results in the query plan holding two copies of the `ScopedCollectionMetadata`. As initially conceived, the design is for a query plan to hold a single `ScopedCollectionMetadata` inside of the `ShardFilterStage`.